### DISCUSSION TITLE
What tool is used to create RLE fonts?

### DISCUSSION THREAD

**Discussion by Murloc992** [Bodmer/TFT_eSPI#855] [Q&A - Fonts] [answered]
Hello,

I was wondering what tool should be used to create my own compatible RLE compressed font source files?

I was using LCD-Image-Converter for a long time but it doesn't seem like the format in use here.

I would really love to know what tool has been used. :)

**Comment by Bodmer** [comment, marked as answer]
A long time ago I write an Arduino sketch that created the fonts from bitmaps, this then needed some manual editting. The bitmaps were generated by a software tool from crystalfontz which does not seem to exist anymore. Unfortunately during a PC upgrade the sketch and tool was lost. Writing again from scratch is on my todo list but is not a priority due to paid for work at the moment.

FYI a character is drawn as a raster scan within a box size of the characters width and global font height. The RLE format is a byte, the top bit is a flag, the least significant 7 bits are a 7 bit number (0-127). If the flag is 1 then foreground pixels are plotted, if it is 0 the background pixels are plotted. The number of pixels to plot is the 7 bit number +1.

So, for example, the encoded sequence "0x81, 0x1F" means plot 2 pixels in foreground then 32 pixels in background.

You could manually generate a single character using some squared paper and a pencil and replace the one in an existing file.

**Reply by JBad13** [reply]
i've written a python script that generates the sequence from a bitmap image of a character
i'm new to github, can i post it here?

**Reply by topoll113** [reply]
@Bodmer Could you explain better or show piece of paper & pencil method? Or maybe you found time to make that script again? I really like your library and fonts that are included but for my project  I need Font 7 smaller . I added that font as GFX but Im not happy about flickering  on the screen + usage of your RLE fonts its way easier . 
Thank you in advance.

**Reply by Bodmer** [reply]
Unfortunately it is quite a complicated process so it requires a deep understanding of the existing font files. This means it is not easily described.

**Reply by topoll113** [reply]
Okay . Ill try figure it out anyway or find different solution . Thank you for your answer.

**Reply by JBad13** [reply]
i wrote this python script for generating rle arrays to be written in .c file
you need a single .bmp file for character
i wrote it 2 years ago, don't remember exactly how i did it, but hope it helps :)

```
from PIL import Image
import sys
import os

def convert_img_to_rle(filename):

    caratteri_speciali = {'spazio'  : ' ',
                          'punto'   : '.',
                          'duepunti': ':',
                          'meno'    : '-'}
    
    def calc_value(px, count):
        palette = img.getpalette()
        if (palette):               # leggo il colore dalla palette
            color = palette[px * 3 : px * 3 + 3]
        else:
            color = [255, 255, 255] # nella bitmap dello spazio non c'e' palette, considero tutto bianco
        
        if (color == [0, 0, 0]):    # pixel neri = colore fg
            value = 0b10000000
        else:                       # pixel colorati = colore bg
            value = 0b00000000
        value += count
        return hex(value)

    if (filename[:-4] in caratteri_speciali):
        carattere = caratteri_speciali[filename[:-4]]   # se nomefile e' nei caratteri speciali, prendo il carattere indicato
    else:
        carattere = filename[0]                         # altrimenti prendo nomefile come carattere
    
    res = '// carattere \'' + carattere + '\', ' + filename + '\n'
    res += 'PROGMEM const unsigned char chr_f72_' + hex(ord(carattere))[2:] + '[] = \n{\n'
    
    img = Image.open(filename)
    data = list(img.getdata())

    for i in range(len(data)):
        if (i == 0):                                            # primo pixel
            last_px = data[i]                                   # pixel on/off
            count = 0                                           # inizia conteggio
        else:
            if ((data[i] == last_px) and (count != 127)):       # stesso stato pixel e conteggio < 127
                count += 1                                      # continuo a contare
            else:
                res += calc_value(last_px, count) + ', '        # altrimenti scrivo e ricomincio conteggio
                last_px = data[i]
                count = 0

    if (count != 0):                                            # scrivo gli ultimi bit
        res += calc_value(last_px, count) + ', '

    res = res[:-2] + "\n};"
   
    return res

for f in os.listdir('.'):
    if (f[-4:].lower() == '.bmp'):
        print convert_img_to_rle(f)

```

**Reply by topoll113** [reply]
@JBad13 Thank you so much for your script. It helped  but I think I’m doing something wrong somewhere . 
![ABDDFB86-202A-40F4-A31C-5FDDBD6AF699](https://user-images.githubusercontent.com/28602900/224174867-26e0044f-98b7-4125-8cf2-c315f06e4542.jpeg) I created bitmap with “0” and replaced original 0 but looks like bytes are in wrong spot or something. Well looks like I need spend more time on it and hopefully solution will appear.

**Reply by Bodmer** [reply]
The width and height of the font needs to be changed.
